File: engine/cssv/CssvRegistry.js
/**
 * CSSV registry — browser + Node hosts register artifacts, transitions, frames.
 * Canonical constitutional ledger API (JS).
 */

import {
  appendNdjson,
  ledgerPaths,
  loadArtifacts,
  saveArtifacts,
} from "./ledger.js";

export class BrowserCssvHost {
  constructor() {
    this.hostId = "browser";
    this.hostVersion = "1.0";
  }
}

/**
 * In-memory registry with optional Node persistence to cssv/ ledger files.
 */
export class CssvRegistry {
  /**
   * @param {{ host?: BrowserCssvHost, persist?: boolean, root?: string }} opts
   */
  constructor(opts = {}) {
    this.host = opts.host ?? new BrowserCssvHost();
    this.persist = opts.persist ?? false;
    this.root = opts.root;
    this.paths = ledgerPaths(this.root);
    this.artifacts = [];
    this.transitions = [];
    this.frames = [];
    this._frameIndex = 0;
    this._stateId = "state-0000";
  }

  registerArtifact(artifact) {
    const record = {
      type: "artifact",
      id: artifact.id,
      artifactType: artifact.artifactType ?? artifact.type,
      host: artifact.hostId ?? this.host.hostId,
      timestamp: Date.now() / 1000,
      payload: {
        dto: artifact.dto ?? "unknown",
        data: artifact.payload ?? artifact.data ?? {},
      },
    };
    this.artifacts.push(record);
    if (this.persist) this._persistArtifact(record);
    return record;
  }

  registerTransition(transition) {
    const record = {
      type: "transition",
      id: transition.id ?? `tx-${this.transitions.length + 1}`,
      from: transition.fromStateId ?? this._stateId,
      to: transition.toStateId ?? `state-${this.transitions.length + 1}`,
      intent: transition.intent ?? {},
      authority: transition.authority ?? this.host.hostId,
      evidence: normalizeEvidence(transition.evidence),
      decision: transition.decision ?? { allowed: true, reasons: [] },
      host: transition.hostId ?? this.host.hostId,
      timestamp: transition.timeSeconds ?? Date.now() / 1000,
    };
    this._stateId = record.to;
    this.transitions.push(record);
    if (this.persist) appendNdjson(this.paths.transitions, record).catch(() => {});
    return record;
  }

  registerFrame(frame) {
    const record = {
      type: "frame",
      intent: frame.intentId ?? frame.intent ?? null,
      timeline: frame.timelineId ?? frame.timeline ?? null,
      world: frame.worldId ?? frame.world ?? null,
      host: frame.hostId ?? this.host.hostId,
      timestamp: frame.timeSeconds ?? Date.now() / 1000,
      params: frame.parameters ?? frame.params ?? {},
      frameIndex: this._frameIndex++,
    };
    this.frames.push(record);
    if (this.persist) appendNdjson(this.paths.frames, record).catch(() => {});
    return record;
  }

  exportSnapshot() {
    return {
      artifacts: this.artifacts.slice(),
      transitions: this.transitions.slice(),
      frames: this.frames.slice(),
    };
  }

  /**
   * Export ledger snapshot for host download.
   * Browser hosts should call downloadBlob / save dialog — this method no longer touches `document`.
   * @returns {{ filename: string, snapshot: object, json: string, bytes: number }}
   */
  exportLedgerDownload(filename = "cssv-ledger") {
    const stamp = new Date().toISOString().replace(/[:.]/g, "-");
    const name = `${filename}-${stamp}.json`;
    const snapshot = this.exportSnapshot();
    const json = JSON.stringify(snapshot, null, 2);
    return { filename: name, snapshot, json, bytes: json.length };
  }

  /**
   * @deprecated Prefer exportLedgerDownload + host downloadBlob. Kept as alias for browser callers.
   */
  downloadLedger(filename = "cssv-ledger") {
    const pack = this.exportLedgerDownload(filename);
    if (typeof this.host?.downloadTextFile === "function") {
      this.host.downloadTextFile(pack.filename, pack.json);
      return { filename: pack.filename, bytes: pack.bytes };
    }
    // No DOM here — return pack for the caller (js/app.js) to download.
    return pack;
  }

  /**
   * POST session snapshot to CSSV server (e.g. http://localhost:3000/ingest).
   * @returns {Promise<{ ok: boolean, ingested?: object, error?: string }>}
   */
  async syncToServer(ingestUrl) {
    if (!ingestUrl) return { ok: false, error: "No ingest URL" };
    const snapshot = this.exportSnapshot();
    try {
      const res = await fetch(ingestUrl, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(snapshot),
      });
      const data = await res.json();
      if (!res.ok) return { ok: false, error: data.error ?? res.statusText };
      return { ok: true, ingested: data.ingested };
    } catch (err) {
      return {
        ok: false,
        error: err instanceof Error ? err.message : String(err),
      };
    }
  }

  async _persistArtifact(record) {
    const existing = await loadArtifacts(this.paths.artifacts);
    const idx = existing.findIndex((a) => a.id === record.id);
    if (idx >= 0) existing[idx] = record;
    else existing.push(record);
    await saveArtifacts(this.paths.artifacts, existing);
  }
}

function normalizeEvidence(evidence) {
  if (!evidence) return [];
  if (Array.isArray(evidence)) return evidence;
  if (Array.isArray(evidence.evidenceIds)) return evidence.evidenceIds;
  if (Array.isArray(evidence.items))
    return evidence.items.map((i) => i.id).filter(Boolean);
  if (evidence.id) return [evidence.id];
  return [];
}

File: unity/GovernedUnityProject/Assets/Engine/Shared/CSSV/CssvRegistry.cs
// CSSV registry — Unity host mirror of engine/cssv/CssvRegistry.js

using System;
using System.Collections.Generic;
using System.Text;
using SovereignX.CIEMS.Engine.Runtime;
using UnityEngine;
using UnityEngine.Networking;

namespace SovereignX.CIEMS.Engine.CSSV
{
    public sealed class UnityCssvHost
    {
        public string HostId { get; set; } = "unity";
        public string HostVersion { get; set; } = "1.0";
    }

    public sealed class CssvArtifactRecord
    {
        public string Id;
        public string ArtifactType;
        public object Payload;
    }

    public sealed class CssvTransitionRecord
    {
        public string Id;
        public string FromStateId;
        public string ToStateId;
        public IntentRecord Intent;
        public string Authority;
        public EvidenceBundle Evidence;
        public Decision Decision;
        public double TimeSeconds;
    }

    /// <summary>In-memory constitutional ledger; optional sync to CSSV server via UnityWebRequest.</summary>
    public sealed class CssvRegistry
    {
        public static CssvRegistry Instance { get; private set; }

        public static CssvRegistry EnsureInstance(UnityCssvHost host = null)
        {
            if (Instance == null)
                Instance = new CssvRegistry(host ?? new UnityCssvHost());
            return Instance;
        }

        readonly UnityCssvHost _host;
        string _stateId = "state-0000";
        int _frameIndex;

        public List<object> Artifacts { get; } = new List<object>();
        public List<object> Transitions { get; } = new List<object>();
        public List<object> Frames { get; } = new List<object>();

        public CssvRegistry(UnityCssvHost host)
        {
            _host = host ?? new UnityCssvHost();
            Instance = this;
        }

        public void RegisterArtifact(CssvArtifactRecord artifact)
        {
            if (artifact == null || string.IsNullOrEmpty(artifact.Id)) return;
            var record = new Dictionary<string, object>
            {
                ["type"] = "artifact",
                ["id"] = artifact.Id,
                ["artifactType"] = artifact.ArtifactType ?? "unknown",
                ["host"] = _host.HostId,
                ["timestamp"] = DateTimeOffset.UtcNow.ToUnixTimeSeconds(),
                ["payload"] = artifact.Payload ?? new Dictionary<string, object>(),
            };
            Artifacts.Add(record);
        }

        public void RegisterTransition(CssvTransitionRecord transition)
        {
            if (transition == null) return;
            var toState = transition.ToStateId ?? $"state-{Transitions.Count + 1}";
            var record = new Dictionary<string, object>
            {
                ["type"] = "transition",
                ["id"] = transition.Id ?? $"tx-{Transitions.Count + 1}",
                ["from"] = transition.FromStateId ?? _stateId,
                ["to"] = toState,
                ["intent"] = SerializeIntent(transition.Intent),
                ["authority"] = transition.Authority ?? _host.HostId,
                ["evidence"] = NormalizeEvidence(transition.Evidence),
                ["decision"] = SerializeDecision(transition.Decision),
                ["host"] = _host.HostId,
                ["timestamp"] = transition.TimeSeconds > 0
                    ? transition.TimeSeconds
                    : DateTimeOffset.UtcNow.ToUnixTimeSeconds(),
            };
            _stateId = toState;
            Transitions.Add(record);
        }

        public void RegisterFrame(FrameProvenance frame)
        {
            var record = new Dictionary<string, object>
            {
                ["type"] = "frame",
                ["intent"] = frame.IntentId,
                ["timeline"] = frame.TimelineId,
                ["world"] = frame.WorldId,
                ["host"] = _host.HostId,
                ["timestamp"] = frame.TimeSeconds,
                ["params"] = frame.Parameters ?? new Dictionary<string, double>(),
                ["frameIndex"] = _frameIndex++,
            };
            Frames.Add(record);
        }

        public Dictionary<string, object> ExportSnapshot()
        {
            return new Dictionary<string, object>
            {
                ["artifacts"] = new List<object>(Artifacts),
                ["transitions"] = new List<object>(Transitions),
                ["frames"] = new List<object>(Frames),
            };
        }

        public string ExportJson()
        {
            var snap = ExportSnapshot();
            return MiniJson.Serialize(snap);
        }

        public void SyncToServer(MonoBehaviour runner, string ingestUrl, Action<bool, string> onComplete = null)
        {
            if (runner == null || string.IsNullOrEmpty(ingestUrl))
            {
                onComplete?.Invoke(false, "No runner or ingest URL");
                return;
            }
            runner.StartCoroutine(PostIngest(ingestUrl, onComplete));
        }

        System.Collections.IEnumerator PostIngest(string ingestUrl, Action<bool, string> onComplete)
        {
            var body = ExportJson();
            using var req = new UnityWebRequest(ingestUrl, "POST");
            var bytes = Encoding.UTF8.GetBytes(body);
            req.uploadHandler = new UploadHandlerRaw(bytes);
            req.downloadHandler = new DownloadHandlerBuffer();
            req.SetRequestHeader("Content-Type", "application/json");
            yield return req.SendWebRequest();

#if UNITY_2020_2_OR_NEWER
            var ok = req.result == UnityWebRequest.Result.Success;
#else
            var ok = !req.isNetworkError && !req.isHttpError;
#endif
            onComplete?.Invoke(ok, ok ? req.downloadHandler.text : req.error);
        }

        static Dictionary<string, object> SerializeIntent(IntentRecord intent)
        {
            if (intent == null) return new Dictionary<string, object>();
            return new Dictionary<string, object>
            {
                ["id"] = intent.Id,
                ["type"] = intent.Type ?? intent.Kind,
                ["actor"] = intent.Actor,
                ["world"] = intent.World,
                ["timeline"] = intent.Timeline,
            };
        }

        static Dictionary<string, object> SerializeDecision(Decision decision)
        {
            if (decision == null) return new Dictionary<string, object> { ["allowed"] = false };
            return new Dictionary<string, object>
            {
                ["allowed"] = decision.Ok,
                ["verdict"] = decision.Verdict,
                ["reason"] = decision.Reason,
                ["violations"] = decision.Violations ?? new List<string>(),
            };
        }

        static List<string> NormalizeEvidence(EvidenceBundle evidence)
        {
            var ids = new List<string>();
            if (evidence == null) return ids;
            if (!string.IsNullOrEmpty(evidence.Id)) ids.Add(evidence.Id);
            if (evidence.Fields != null &&
                evidence.Fields.TryGetValue("evidenceIds", out var list) &&
                list is IEnumerable<string> enumerable)
            {
                foreach (var id in enumerable)
                    if (!string.IsNullOrEmpty(id)) ids.Add(id);
            }
            return ids;
        }
    }

    /// <summary>Minimal JSON writer for CSSV snapshots (no external deps).</summary>
    static class MiniJson
    {
        public static string Serialize(object value) => SerializeValue(value);

        static string SerializeValue(object value)
        {
            if (value == null) return "null";
            if (value is string s) return "\"" + Escape(s) + "\"";
            if (value is bool b) return b ? "true" : "false";
            if (value is int or long or float or double)
                return Convert.ToString(value, System.Globalization.CultureInfo.InvariantCulture);
            if (value is IDictionary<string, object> dict)
            {
                var parts = new List<string>();
                foreach (var kv in dict)
                    parts.Add("\"" + Escape(kv.Key) + "\":" + SerializeValue(kv.Value));
                return "{" + string.Join(",", parts) + "}";
            }
            if (value is IDictionary<string, double> dblDict)
            {
                var parts = new List<string>();
                foreach (var kv in dblDict)
                    parts.Add("\"" + Escape(kv.Key) + "\":" + SerializeValue(kv.Value));
                return "{" + string.Join(",", parts) + "}";
            }
            if (value is IEnumerable<object> list)
            {
                var parts = new List<string>();
                foreach (var item in list)
                    parts.Add(SerializeValue(item));
                return "[" + string.Join(",", parts) + "]";
            }
            if (value is IEnumerable<string> strList)
            {
                var parts = new List<string>();
                foreach (var item in strList)
                    parts.Add(SerializeValue(item));
                return "[" + string.Join(",", parts) + "]";
            }
            if (value is System.Collections.IEnumerable enumerable && value is not string)
            {
                var parts = new List<string>();
                foreach (var item in enumerable)
                    parts.Add(SerializeValue(item));
                return "[" + string.Join(",", parts) + "]";
            }
            return "\"" + Escape(value.ToString()) + "\"";
        }

        static string Escape(string s) =>
            (s ?? "").Replace("\\", "\\\\").Replace("\"", "\\\"").Replace("\n", "\\n");
    }
}
